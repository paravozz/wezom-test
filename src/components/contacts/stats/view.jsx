import React from 'react';
import { Typography, Col, Row, Statistic, Descriptions } from 'antd';
import {
	countGenders,
	getGenderPredominateMessage,
	countNationalities,
	CountryData,
} from 'utils';


const View = (props) => {
	const { contacts, loading } = props;

	if (!contacts?.length || loading) {
		return (
			<Row className="box" />
		);
	}

	function renderNationalitiesStats(_contacts) {
		const natCount = countNationalities(_contacts);

		return (
			<Descriptions title="Nationalities">
				{Object.entries(natCount).map(([key, value]) => (
					<Descriptions.Item key={key} label={CountryData[key].name}>
						{value} contacts
					</Descriptions.Item>
				))}
			</Descriptions>
		);
	}

	const genderStats = countGenders(contacts);
	const genderPredominateMessage = getGenderPredominateMessage(genderStats);

	return (
		<div className="box">
			<Typography.Title level={3}>Statistic</Typography.Title>
			<Row gutter={24}>
				<Col className="_flex-noshrink">
					<Statistic title="Collection size" value={contacts.length} />
				</Col>
				<Col className="_flex-noshrink">
					<Row gutter={12}>
						<Col span={6}>
							<Statistic title="Males" value={genderStats.male} />
						</Col>
						<Col span={6}>
							<Statistic title="Females" value={genderStats.female} />
						</Col>
						<Col span={6}>
							<Statistic title="Indeterminate" value={genderStats.indeterminate} />
						</Col>
						<Col span={24}>
							<Typography.Text mark>
								{genderPredominateMessage}
							</Typography.Text>
						</Col>
					</Row>
				</Col>
			</Row>
			<br />
			<Row>
				{renderNationalitiesStats(contacts)}
			</Row>
		</div>
	);
};


export { View };
