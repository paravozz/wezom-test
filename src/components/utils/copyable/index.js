export { default as Copyable } from './view';
