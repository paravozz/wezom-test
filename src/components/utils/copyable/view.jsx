import React from 'react';

import { CopyToClipboard } from 'react-copy-to-clipboard';
import { CopyOutlined, CheckOutlined } from '@ant-design/icons';
import { Tooltip } from 'antd';

import './style.scss';


export default ({ children, text }) => {
	const [isCopied, setIsCopied] = React.useState(false);


	React.useEffect(() => {
		if (isCopied) {
			setTimeout(() => {
				setIsCopied(false);
			}, 3000);
		}
	}, [isCopied]);


	return (
		<div className="copyable--container">
			<CopyToClipboard onCopy={() => setIsCopied(true)} text={text}>
				<Tooltip title={isCopied ? 'Copied' : 'Copy'}>
					{
						isCopied
							? <CheckOutlined style={{ color: '#52c41a' }} />
							: <CopyOutlined style={{ color: '#1890ff' }} />
					}
				</Tooltip>
			</CopyToClipboard>
			{children}
		</div>
	)
};
