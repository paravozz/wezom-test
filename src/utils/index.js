export {
	provide,
	provideWithRandomError,
} from './provide';
export {
	randomIntegerInRange,
} from './random';
export { redirect } from './redirect';
export {
	getFromStorage,
	setToStorage,
} from './local-storage-api';
export {
	countGenders,
	getGenderPredominateMessage,
	countNationalities,
} from './contacts-stats';
export { CountryData } from './countries';
export {
	getFullName,
	getGenderName,
	mapGenderToItsName,
	paginateCollection
} from './contacts';
