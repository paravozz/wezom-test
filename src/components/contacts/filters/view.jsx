import React from 'react';
import { Row, Input, Select, Col, Checkbox, Button } from 'antd';
import { CloseOutlined } from '@ant-design/icons';
import { CountryData } from 'utils';
import { getGenderName } from 'utils';


const View = (props) => {
	const {
		loading,
		filters,
		filterContacts,
	} = props;

	return (
		<div className="box">
			<Row align="middle">
				<Col className="_flex-grow">
					<Row gutter={[6, 6]} align="middle">
						<Col xs={24} lg={11}>
							<Row>
								<Input.Search
									loading={loading}
									placeholder="Search by full name"
									size="large"
									onChange={e => filterContacts({ ...filters, name: e.target.value })}
									value={filters.name}
								/>
							</Row>
						</Col>
						<Col xs={24} sm={10} lg={4}>
							<Row>
								<Select
									id="genderSelect"
									loading={loading}
									style={{ width: '100%' }}
									allowClear
									placeholder="Gender"
									size="large"
									onChange={value => filterContacts({ ...filters, gender: value })}
									value={filters.gender || undefined}
								>
									{['male', 'female', 'indeterminate'].map(k => (
										<Select.Option key={k} value={k}>
											{getGenderName(k)}
										</Select.Option>
									))}
								</Select>
							</Row>
						</Col>
						<Col xs={24} sm={14} lg={5}>
							<Row>
								<Select
									id="natSelect"
									loading={loading}
									style={{ width: '100%' }}
									mode="multiple"
									maxTagCount={2}
									allowClear
									size="large"
									placeholder="Nationality"
									onChange={value => filterContacts({ ...filters, nat: value })}
									value={filters.nat}
								>
									{Object.entries(CountryData).map(([key, data]) => (
										<Select.Option key={key} value={key}>{data.name}</Select.Option>
									))}
								</Select>
							</Row>
						</Col>
						<Col xs={24} sm={6} lg={4}>
							<Row>
								<Checkbox
									disabled={loading}
									onChange={(e) => filterContacts({ ...filters, author: e.target.checked })}
								>
									I am creator
								</Checkbox>
							</Row>
						</Col>
					</Row>
				</Col>
				<Col className="_flex-noshrink">
					<Row>
						<Button
							disabled={loading}
							onClick={() => filterContacts(null)}
							type="link"
						>
							<CloseOutlined /> Clear
						</Button>
					</Row>
				</Col>
			</Row>
		</div>
	);
};


export { View };
