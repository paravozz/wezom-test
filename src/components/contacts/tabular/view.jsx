import React from 'react';
import { Table, Typography, Tag, Avatar, Empty } from 'antd';
import { UserOutlined } from '@ant-design/icons';
import moment from 'moment';
import { CountryData, getFullName } from 'utils';
import { Copyable } from 'components/utils';


const View = (props) => {
	const { contacts, loading } = props;

	if (!contacts) {
		return (
			<>
				<Empty image={Empty.PRESENTED_IMAGE_SIMPLE} />
			</>
		)
	}

	function renderFullName({ title, first, last }) {
		return getFullName(title, first, last);
	}

	function renderAvatar(pictureUrl) {
		return (
			<Avatar size={40} icon={<UserOutlined />} src={pictureUrl} />
		);
	}

	function renderBirthday({ date, age }) {
		return (
			<div>
				{moment(date).format('dddd, M/D/YYYY, H:m:s A')}
				<br />
				{age} years
			</div>
		);
	}

	function renderEmail(email) {
		return (
			<Copyable text={email}>
				<Typography.Link ellipsis href={`mailto:${email}`}>
					{email}
				</Typography.Link>
			</Copyable>
		);
	}

	function renderPhone(phone) {
		return (
			<Copyable text={phone}>
				<Typography.Link ellipsis href={`tel:${phone}`}>
					{phone}
				</Typography.Link>
			</Copyable>
		);
	}

	function renderLocation({ country, street, city, state, postcode }) {
		return (
			<Copyable
				text={(
					`[${country}] `
					+ `${street.number} ${street.name}, `
					+ `${city}, ${state} ${postcode}`
				)}
			>
				<Typography.Paragraph ellipsis={{ rows: 3 }}>
					<strong>/{country}/</strong>
					<br />
					{(
						`${street.number}, ${street.name}`
						+ `${city}, ${state} ${postcode}`
					)}
				</Typography.Paragraph>
			</Copyable>
		);
	}

	function renderNationality(nat) {
		return (
			<Tag color={CountryData[nat].color}>{CountryData[nat].name}</Tag>
		)
	}

	const dataSource = (
		contacts
			.map((contact, index) => ({
				key: index,
				avatar: renderAvatar(contact.picture.thumbnail),
				fullName: renderFullName(contact.name),
				birthday: renderBirthday(contact.dob),
				email: renderEmail(contact.email),
				phone: renderPhone(contact.phone),
				location: renderLocation(contact.location),
				nationality: renderNationality(contact.nat),
			}))
	);

	const columns = [
		{
			title: 'Avatar',
			dataIndex: 'avatar',
			key: 'avatar',
			fixed: 'left',
			width: 80,
		},
		{
			title: 'Full name',
			dataIndex: 'fullName',
			key: 'fullName',
			width: 155,
			sorter: (a, b) => a.fullName.length - b.fullName.length,
			sortDirections: ['descend', 'ascend'],
		},
		{
			title: 'Birthday',
			dataIndex: 'birthday',
			key: 'birthday',
			width: 250,
		},
		{
			title: 'Email',
			dataIndex: 'email',
			key: 'email',
			width: 155,
		},
		{
			title: 'Phone',
			dataIndex: 'phone',
			key: 'phone',
			width: 155,
		},
		{
			title: 'Location',
			dataIndex: 'location',
			key: 'location',
			width: 250,
		},
		{
			title: 'Nationality',
			dataIndex: 'nationality',
			key: 'nationality',
			width: 155,
			align: 'right',
		},
	];


	return (
		<Table
			dataSource={dataSource}
			columns={columns}
			scroll={{ x: '100vw' }}
			loading={loading}
			pagination={{ position: ['bottomCenter'] }}
		/>
	);
};


export { View };
