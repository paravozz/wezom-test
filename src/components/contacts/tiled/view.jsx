import React from 'react';
import { Typography, Tag, Row, Col, Card, Divider, Empty, Spin } from 'antd';
import { CountryData, getFullName, paginateCollection } from 'utils';
import { Copyable } from 'components/utils';


function renderFullName({ title, first, last }) {
	return getFullName(title, first, last);
}

function renderAvatar(pictureUrl) {
	return (
		<img src={pictureUrl} alt="contact-cover" />
	);
}

function renderBirthday({ age }) {
	return `${age} years`
}

function renderEmail(email) {
	return (
		<Copyable text={email}>
			<Typography.Link ellipsis href={`mailto:${email}`}>
				{email}
			</Typography.Link>
		</Copyable>
	);
}

function renderPhone(phone) {
	return (
		<Copyable text={phone}>
			<Typography.Link ellipsis href={`tel:${phone}`}>
				{phone}
			</Typography.Link>
		</Copyable>
	);
}

function renderLocation({ country, street, city, state, postcode }) {
	return (
		<Copyable
			text={(
				`[${country}] `
				+ `${street.number} ${street.name}, `
				+ `${city}, ${state} ${postcode}`
			)}
		>
			<Typography.Paragraph ellipsis={{ rows: 3 }}>
				<strong>/{country}/</strong>
				<br />
				{(
					`${street.number}, ${street.name}`
					+ `${city}, ${state} ${postcode}`
				)}
			</Typography.Paragraph>
		</Copyable>
	);
}

function renderNationality(nat) {
	return (
		<Tag color={CountryData[nat].color}>{CountryData[nat].name}</Tag>
	)
}

export function transformContacts(pagination, contacts) {
	let baseCollection = (
		pagination
			? paginateCollection(contacts, pagination.offset, pagination.limit)
			: contacts
	);

	return baseCollection.map((contact, index) => ({
		key: index,
		avatar: renderAvatar(contact.picture.large),
		fullName: renderFullName(contact.name),
		birthday: renderBirthday(contact.dob),
		email: renderEmail(contact.email),
		phone: renderPhone(contact.phone),
		location: renderLocation(contact.location),
		nationality: renderNationality(contact.nat),
	}));
}


const View = (props) => {
	const { contacts, loading, pagination } = props;

	if (!contacts?.length || loading) {
		return (
			<>
				{loading && <Spin />}
				<Empty image={Empty.PRESENTED_IMAGE_SIMPLE} />
			</>
		)
	}

	return (
		<Row gutter={[8, 8]}>
			{transformContacts(pagination, contacts).map(contact => (
				<div key={contact.key} style={{ width: '33.3333%', maxWidth: '33.3333%' }}>
					<Col>
						<Card
							hoverable
							cover={contact.avatar}
						>
							<Card.Meta
								title={(
									<Typography.Text>
										<Typography.Text>{contact.fullName}</Typography.Text>
										&nbsp;
										<Typography.Text type="secondary">
											({contact.birthday})
										</Typography.Text>
									</Typography.Text>
								)}
								description={(
									<>
										{contact.email}
										{contact.phone}
										{contact.location}
										<Divider/>
										{contact.nationality}
									</>
								)}
							/>
						</Card>
					</Col>
				</div>
			))}
		</Row>
	);
};


export { View };
