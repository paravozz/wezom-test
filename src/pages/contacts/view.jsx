import React from 'react';

import { Radio, Row, Col, Typography, Pagination, Tooltip, Button } from 'antd';
import { DatabaseOutlined, AppstoreOutlined, ReloadOutlined } from '@ant-design/icons';

import { getFromStorage, randomIntegerInRange, setToStorage } from 'utils';
import {
	TabularContacts,
	StatsForContacts,
	FiltersForContacts,
} from 'components';
import { TiledContacts } from '../../components/contacts/tiled';


const __VIEW_MODE_KEY__ = 'app__view_mode';


const View = (props) => {
	const {
		totalCount,
		fetchContacts,
		pagination,
		setPagination,
		contacts,
	} = props;

	const [viewMode, setViewMode] = React.useState(getFromStorage(__VIEW_MODE_KEY__) || 'tabular');

	function fetchContactsRemote() {
		fetchContacts(totalCount || randomIntegerInRange(1, 1000));
	}


	React.useEffect(() => {
		fetchContactsRemote();
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, []);

	React.useEffect(() => {
		setToStorage(__VIEW_MODE_KEY__, viewMode);
	}, [viewMode]);


	const radioOptions = [
		{ label: <DatabaseOutlined />, value: 'tabular' },
		{ label: <AppstoreOutlined />, value: 'tiled' },
	];

	const componentForViewMode = {
		tiled: (
			<>
				<TiledContacts />
				<Pagination
					onChange={(page, pageSize) => (
						setPagination({
							offset: pageSize * (page - 1),
							limit: pageSize,
						})
					)}
					current={pagination.offset / pagination.limit + 1}
					pageSize={pagination.limit}
					pageSizeOptions={[6, 12, 24, 36, 48]}
					total={contacts.length}
				/>
			</>
		),
		tabular: <TabularContacts />,
	};


	return (
		<div className={'page page--contacts'}>
			<Row justify="space-between">
				<Typography.Title level={2}>Contacts</Typography.Title>
				<Col>
					<Tooltip title="Reload data">
				  		<Button
							type="dashed"
							shape="circle"
							icon={<ReloadOutlined />}
							onClick={() => fetchContactsRemote()}
						/>
					</Tooltip>
					<Radio.Group
						options={radioOptions}
						onChange={(e) => setViewMode(e.target.value)}
						value={viewMode}
						optionType="button"
						buttonStyle="solid"
					/>
				</Col>
			</Row>
			<FiltersForContacts />
			{componentForViewMode[viewMode]}
			<StatsForContacts />
		</div>
	);
};


export { View };
