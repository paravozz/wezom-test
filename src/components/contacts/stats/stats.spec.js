/* eslint-disable */
import React from 'react';
import { View as Stats } from './view.jsx';
import { Row } from 'antd';
import { countNationalities, countGenders, getGenderPredominateMessage } from 'utils';


jest.mock('utils');
const setUp = (props) => shallow(<Stats {...props} />);
const setUpRendered = (props) => render(<Stats {...props} />);

describe('Stats', () => {
	let component;

	beforeEach(() => {
		jest.resetModules();
		component = setUp();

		countGenders.mockReturnValue({
			male: 1,
			female: 2,
			indeterminate: 0,
		});
		getGenderPredominateMessage.mockReturnValue('Female predominate');
		countNationalities.mockReturnValue({
			NZ: 1,
			GB: 1,
			AU: 1,
		});
	});

	it('should render without crashing', () => {
		component = setUpRendered({ contacts });
		expect(component).toMatchSnapshot();
	});

	it('should render without redux props', () => {
		expect(component)
	});

	it('should contain Row component if no props provided', () => {
		expect((component.find(Row))?.length).toBe(1);
	});

	it('should call all calculation functions in render cycle', () => {
		component = setUp({ contacts });

		expect(countGenders).toBeCalled();
		expect(countGenders).toBeCalledWith(contacts);

		expect(getGenderPredominateMessage).toBeCalled();
		expect(getGenderPredominateMessage).toBeCalledWith(countGenders());

		expect(countNationalities).toBeCalled();
		expect(countNationalities).toBeCalledWith(contacts);
	});
});
