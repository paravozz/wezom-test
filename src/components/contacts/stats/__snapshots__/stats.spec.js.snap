// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats should render without crashing 1`] = `
<div
  class="box"
>
  <h3
    class="ant-typography"
  >
    Statistic
  </h3>
  <div
    class="ant-row"
    style="margin-left:-12px;margin-right:-12px"
  >
    <div
      class="ant-col _flex-noshrink"
      style="padding-left:12px;padding-right:12px"
    >
      <div
        class="ant-statistic"
      >
        <div
          class="ant-statistic-title"
        >
          Collection size
        </div>
        <div
          class="ant-statistic-content"
        >
          <span
            class="ant-statistic-content-value"
          >
            <span
              class="ant-statistic-content-value-int"
            >
              3
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="ant-col _flex-noshrink"
      style="padding-left:12px;padding-right:12px"
    >
      <div
        class="ant-row"
        style="margin-left:-6px;margin-right:-6px"
      >
        <div
          class="ant-col ant-col-6"
          style="padding-left:6px;padding-right:6px"
        >
          <div
            class="ant-statistic"
          >
            <div
              class="ant-statistic-title"
            >
              Males
            </div>
            <div
              class="ant-statistic-content"
            >
              <span
                class="ant-statistic-content-value"
              >
                <span
                  class="ant-statistic-content-value-int"
                >
                  1
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-6"
          style="padding-left:6px;padding-right:6px"
        >
          <div
            class="ant-statistic"
          >
            <div
              class="ant-statistic-title"
            >
              Females
            </div>
            <div
              class="ant-statistic-content"
            >
              <span
                class="ant-statistic-content-value"
              >
                <span
                  class="ant-statistic-content-value-int"
                >
                  2
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-6"
          style="padding-left:6px;padding-right:6px"
        >
          <div
            class="ant-statistic"
          >
            <div
              class="ant-statistic-title"
            >
              Indeterminate
            </div>
            <div
              class="ant-statistic-content"
            >
              <span
                class="ant-statistic-content-value"
              >
                <span
                  class="ant-statistic-content-value-int"
                >
                  0
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-24"
          style="padding-left:6px;padding-right:6px"
        >
          <span
            class="ant-typography"
          >
            <mark>
              Female predominate
            </mark>
          </span>
        </div>
      </div>
    </div>
  </div>
  <br />
  <div
    class="ant-row"
  >
    <div
      class="ant-descriptions"
    >
      <div
        class="ant-descriptions-header"
      >
        <div
          class="ant-descriptions-title"
        >
          Nationalities
        </div>
      </div>
      <div
        class="ant-descriptions-view"
      >
        <table>
          <tbody>
            <tr
              class="ant-descriptions-row"
            >
              <td
                class="ant-descriptions-item"
                colspan="1"
              >
                <span
                  class="ant-descriptions-item-label"
                >
                  New Zealander
                </span>
                <span
                  class="ant-descriptions-item-content"
                >
                  1 contacts
                </span>
              </td>
              <td
                class="ant-descriptions-item"
                colspan="1"
              >
                <span
                  class="ant-descriptions-item-label"
                >
                  British
                </span>
                <span
                  class="ant-descriptions-item-content"
                >
                  1 contacts
                </span>
              </td>
              <td
                class="ant-descriptions-item"
                colspan="1"
              >
                <span
                  class="ant-descriptions-item-label"
                >
                  Australian
                </span>
                <span
                  class="ant-descriptions-item-content"
                >
                  1 contacts
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
